Get cartographer working under ROS2

clalancette opened the pull request:
This PR is the work I have done so far to get cartographer_ros working under ROS2.  This is the bulk of the work that needs to be done, and with this code, the cartographer_ros node builds and starts up.  I haven't been able to test it out fully yet because I still need to do some additional work to translate depth camera data to point cloud data.  I'm working on that now, but I figured I would start this review in the meantime.

connects to ros2/ros2#342

wjwwood commented:
What's the state on this? Should I start a fresh review or is @mikaelarguedas still in the process of reviewing it? Are there outstanding comments that need addressing?

mikaelarguedas commented:
You can go for it, I made a few comments only on what prevented it from building on my machine. I haven't reviewed or tested the code.
Note: we need forks of ceres and cartographer if we want to compile them (and this) without catkin.

wjwwood commented:
> Note: we need forks of ceres and cartographer if we want to compile them (and this) without catkin.

Neither ceres nor cartographer uses catkin afaik. They both have `package.xml` upstream (version 2), but they are plain cmake packages and not catkin packages.

Why would this fork need catkin? Would that be part of the port?

I can give it a go reviewing it in any case.

mikaelarguedas commented:
> > Note: we need forks of ceres and cartographer if we want to compile them (and this) without catkin.

> Neither ceres nor cartographer uses catkin afaik. They both have package.xml upstream (version 2), but they are plain cmake packages and not catkin packages.

Saying this because I noticed that @clalancette was using https://github.com/asirobots/cartographer and https://github.com/asirobots/ceres-solver for building. I figured the reason was because it didn't work out of the box with a package.xml that buildtool_depend on catkin. But maybe the reason was something else

wjwwood commented on cartographer_ros/cartographer_ros/sensor_bridge.cc:
```diff
@@ -20,6 +20,77 @@
+  void toPCL(const std_msgs::msg::Header &header, pcl::PCLHeader &pcl_header)
+  {
+    toPCL(header.stamp, pcl_header.stamp);
+    // FIXME: Seq doesn't exist anymore
```
Please consider putting your github username in parenthesis like this: `// FIXME(clalancette): ...`. The idea is that we can easily know who to ask about it if the comment is no longer clear, without resorting to using `git blame`. If you don't see a need for it then that's totally fine, but I personally think it's a good habit to get into.

wjwwood replied:
Also, we use `TODO` more often, but `FIXME` is just as searchable.

clalancette replied:
Yeah, that's fine, I'll switch to TODO(clalancette); I just have to get used to that style.

wjwwood commented on cartographer_ros/cartographer_ros/tf_bridge.cc:
```diff
@@ -30,23 +30,30 @@ TfBridge::TfBridge(const string& tracking_frame,
 std::unique_ptr<::cartographer::transform::Rigid3d> TfBridge::LookupToTracking(
     const ::cartographer::common::Time time, const string& frame_id) const {
-  ::ros::Duration timeout(lookup_transform_timeout_sec_);
+  tf2::Duration timeout(lookup_transform_timeout_sec_ * 1000000000.0);
```
If you're going to use a float anyways, maybe the `1e9` syntax would be easier to read.

clalancette replied:
Sure, that's fine.

clalancette commented on cartographer_ros/CMakeLists.txt:
```diff
@@ -12,42 +12,38 @@
-google_initialize_cartographer_project()
-google_enable_testing()
+if(NOT WIN32)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -Wall -Wextra -ggdb3")
```
Yes, absolutely.

clalancette commented on cartographer_ros/cartographer_ros/CMakeLists.txt:
```diff
@@ -12,35 +12,50 @@
+  node_main.cc)
+target_include_directories(cartographer_node SYSTEM PUBLIC ${LUA_INCLUDE_DIR})
+target_link_libraries(cartographer_node ${PROJECT_NAME})
+ament_target_dependencies(cartographer_node "rclcpp" "sensor_msgs" "nav_msgs" "tf2_ros" "cartographer_ros_msgs" "tf2" "tf2_msgs" "console_bridge")
```
Oh yeah, good point.  Will fix.

clalancette commented on cartographer_ros/cartographer_ros/assets_writer_main.cc:
```diff
@@ -70,6 +70,7 @@ namespace carto = ::cartographer;
     const sensor_msgs::PointCloud2::ConstPtr& message) {
   pcl::PointCloud<pcl::PointXYZ> pcl_point_cloud;
+  fromROSMsg(*message, pcl_point_cloud);
   pcl::fromROSMsg(*message, pcl_point_cloud);
```
Yeah, silliness.  This was part of my original porting effort, but it turns out that for "live" cartographer, you don't need this file, and it is never compiled.  In point of fact, we can't port this file to ROS2 until we have a story around rosbags.  So I'll just drop this change.

clalancette commented on cartographer_ros/cartographer_ros/map_builder_bridge.cc:
```diff
@@ -64,31 +65,31 @@ void MapBuilderBridge::WriteAssets(const string& stem) {
-    cartographer_ros_msgs::SubmapQuery::Response& response) {
+void MapBuilderBridge::HandleSubmapQuery(
+                                         const std::shared_ptr<::cartographer_ros_msgs::srv::SubmapQuery::Request> request,
+                                         std::shared_ptr<::cartographer_ros_msgs::srv::SubmapQuery::Response> response) {
```
Yeah, sorry.  I still haven't configured emacs totally properly.  I'll fix this.

clalancette commented on cartographer_ros/cartographer_ros/msg_conversion.cc:
```diff
@@ -146,44 +144,50 @@ sensor_msgs::LaserScan ToLaserScan(
-  ::ros::serialization::OStream stream(msg.data.data(), msg.data.size());
+
+  size_t offset = 0;
+  float * const data = reinterpret_cast<float*>(&msg.data[0]);
```
What's happening here is that we need to iterate over the data, one float size at a time, filling in the data.  The code is thus casting us to a float pointer so that we can just do offset++, and move forward the 4 bytes every time.  We can't use a const_cast, both because the data isn't const (we modify it in the below loop), and because we are changing types (from a char pointer to a float pointer).

That all being said, I'm open to other suggestions on how to do this.  This is how I would have done it in C (more-or-less), so I did it that way.  There may be a safer/more C++ way to do it.

wjwwood replied:
Sounds good, I just didn't know what was going on.

clalancette commented on cartographer_ros/cartographer_ros/msg_conversion.h:
```diff
@@ -17,61 +17,63 @@
 #ifndef CARTOGRAPHER_ROS_GOOGLE_CARTOGRAPHER_SRC_MSG_CONVERSION_H_
 #define CARTOGRAPHER_ROS_GOOGLE_CARTOGRAPHER_SRC_MSG_CONVERSION_H_
 
+#include <boost/array.hpp>
```
There are just a few minor uses of boost in cartographer_ros; this is one of them.  I just added the include of the header file to be more correct, since it was having trouble finding the symbols.  While I think removal of boost would be valuable, I think we should do it in upstream (ROS1) cartographer, and then inherit it here.

clalancette commented on cartographer_ros/cartographer_ros/node.cc:
```diff
@@ -62,55 +66,46 @@ Node::~Node() {
+    node_handle_->create_publisher<sensor_msgs::msg::PointCloud2>(
+                                                                  kScanMatchedPointCloudTopic, rmw_qos_profile_default);
+
+  wall_timers_.push_back(node_handle_->create_wall_timer(std::chrono::milliseconds((int)(options_.submap_publish_period_sec*1000)), std::bind(&Node::PublishSubmapList, this)));
```
Yeah, good point about the line length.  I'll fix that, and also change the casting style.

clalancette commented on cartographer_ros/cartographer_ros/node.h:
```diff
@@ -23,13 +23,14 @@
+#include "cartographer_ros_msgs/msg/submap_list.hpp"
+#include "cartographer_ros_msgs/srv/submap_query.hpp"
+#include "cartographer_ros_msgs/msg/trajectory_submap_list.hpp"
+//#include "ros/ros.h"
```
Gone :)

clalancette commented on cartographer_ros/cartographer_ros/node_main.cc:
```diff
@@ -63,103 +63,98 @@ void Run() {
-                  ->HandleImuMessage(kImuTopic, msg);
-            }));
+    imu_subscriber = node.node_handle()->create_subscription<sensor_msgs::msg::Imu>(
+                  kImuTopic,
```
Yes to all of the above :).  Will fix the indentation.

clalancette commented on cartographer_ros/cartographer_ros/ros_log_sink.cc:
```diff
@@ -20,9 +20,15 @@
 
 #include "glog/log_severity.h"
-#include "ros/console.h"
+//#include "ros/console.h"
```
Gone.

clalancette commented:
I've now done most of the updates that William mentioned.

There is one part of this PR that is pretty ugly, and that is the fact that I cut-n-pasted parts of pcl_conversions that I needed into sensor_bridge.cc just to get it to work.  The right thing to do here, undoubtedly, is to port pcl_conversions to ROS2.  I will eventually open up a PR to do that, but I'm not quite there yet.  We can have a discussion on whether we want to wait for that, or if we should just push what I have here for now and do that later.  Thoughts?

mikaelarguedas commented:
I understand that we don't have time to port every package from which we need a single feature. I liked the compromise we adopted everywhere else of cloning and porting what we need (raising not implemented or commenting out what we don't have time to port) rather than copy-pasting code from different ROS packages into another forked one without any reference to where it comes from (same comment as https://github.com/ros2/turtlebot2_demo/pull/11#issuecomment-293739451 or https://github.com/ros2/vision_opencv/pull/1#discussion_r111689338).
If you want to keep this code here for now to move things forward, make sure to add a TODO and open an issue to remind us that this is temporary and should be addressed later.

Also, as discussed offline, +1 to deboostifying when it's a very low hanging fruit, and propagated upstream as you mentioned (same comment as https://github.com/ros2/vision_opencv/pull/1#discussion_r111689380).

clalancette commented:
All right, I believe I have now addressed most of the review comments.  I've also removed the easy-to-remove boost::array, replacing it with std::array.  The only outstanding question revolves around the reinterpret_cast of byte data to float pointers so we can iterator over it.  As mentioned in that comment, I need to do something along those lines, so we either have to stick with the current code or someone needs to tell me how to do it in a nicer C++ way :).

clalancette commented:
CI:

Turtlebot demo [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo&build=26)](http://ci.ros2.org/job/ci_turtlebot-demo/26/)

wjwwood commented on cartographer_ros/cartographer_ros/map_writer.h:
```diff
@@ -19,13 +19,13 @@
 // Writes the given 'occupancy_grid' as 'stem'.pgm and 'stem'.yaml.
 void WriteOccupancyGridToPgmAndYaml(
-    const ::nav_msgs::OccupancyGrid& occupancy_grid, const std::string& stem);
+                                    const ::nav_msgs::msg::OccupancyGrid& occupancy_grid, const std::string& stem);
```
This one still looks weird to me (it doesn't even save line length)?

wjwwood commented on cartographer_ros/cartographer_ros/node.cc:
```diff
@@ -62,55 +66,50 @@ Node::~Node() {
-      ::ros::WallDuration(options_.pose_publish_period_sec),
-      &Node::PublishTrajectoryStates, this));
+    node_handle_->create_publisher<sensor_msgs::msg::PointCloud2>(
+                                                                  kScanMatchedPointCloudTopic, rmw_qos_profile_default);
```
This is another place where the indentation doesn't make sense to me.

clalancette replied:
Yeah, I missed it on the previous comments.  Fixed now.

wjwwood commented on cartographer_ros/cartographer_ros/node_main.cc:
```diff
@@ -63,104 +63,104 @@ void Run() {
+      //if (options.num_point_clouds > 1) {
+      //  topic += "_" + std::to_string(i + 1);
+      //}
+      //point_cloud_topics.push_back(topic);
```
Should this be removed, or uncommented in the future? If the latter, please add a TODO or at least a comment indicating why it is still here and when it could be restored.

clalancette replied:
We'll need to fix it in the future.  This is my workaround for right now, so I'll add a TODO comment to it.

wjwwood commented on cartographer_ros/cartographer_ros/tf_bridge.cc:
```diff
@@ -30,23 +30,33 @@ TfBridge::TfBridge(const string& tracking_frame,
+
+    //return ::cartographer::common::make_unique<
+    //    ::cartographer::transform::Rigid3d>(ToRigid3d(buffer_->lookupTransform(
+    //    tracking_frame_, frame_id, recovered, timeout)));
```
Remove or comment as well.

clalancette replied:
Yeah, another one we'll need to fix.  I'll add a comment.

mikaelarguedas commented on cartographer_ros/cartographer_ros/node.h:
```diff
@@ -23,13 +23,14 @@
+#include "cartographer_ros_msgs/msg/trajectory_submap_list.hpp"
+
 #include "tf2_ros/transform_broadcaster.h"
+#include <rclcpp/rclcpp.hpp>
```
same as above: why this one included with `<` ?

mikaelarguedas commented on cartographer_ros/cartographer_ros/node_main.cc:
```diff
@@ -25,6 +25,8 @@
 #include "cartographer_ros/ros_log_sink.h"
 #include "gflags/gflags.h"
 #include "tf2_ros/transform_listener.h"
+#include <rclcpp/rclcpp.hpp>
```
same as above

mikaelarguedas commented on cartographer_ros/cartographer_ros/node_main.cc:
```diff
@@ -25,6 +25,8 @@
 #include "gflags/gflags.h"
 #include "tf2_ros/transform_listener.h"
+#include <rclcpp/rclcpp.hpp>
+#include "cartographer_ros_msgs/srv/finish_trajectory.hpp"
```
alphabetical

mikaelarguedas commented on cartographer_ros/cartographer_ros/node_main.cc:
```diff
@@ -63,104 +63,109 @@ void Run() {
                   ->sensor_bridge(trajectory_id)
                   ->HandleLaserScanMessage(kLaserScanTopic, msg);
-            }));
+        }, rmw_qos_profile_default);
```
I'd guess you want to use sensor_data profile by befaut during your initial port and iterate on it later

clalancette replied:
Oh, right.  I actually don't use these topics, which is why I didn't care about getting the QoS profile correct for them.  However, I'll make them use the same custom one as I use elsewhere.

mikaelarguedas commented on cartographer_ros/cartographer_ros/offline_node_main.cc:
```diff
@@ -175,7 +175,7 @@ void Run(const std::vector<string>& bag_filenames) {
     bag.open(bag_filename, rosbag::bagmode::Read);
     rosbag::View view(bag);
-    const ::ros::Time begin_time = view.getBeginTime();
+    const ::builtin_interfaces::msg::Time begin_time = view.getBeginTime();
```
should this include builtin_interfaces  and remove ros/time include ? (same everywhere `ros::Time` is replaced by `builtin_interfaces::msg::Time`)

clalancette replied:
Yeah, again, this is a file that isn't compiled, and can't be ported/compiled until we have rosbag support in ROS2.  So we have the same decision to make as for assets_writer_main.cc

mikaelarguedas commented on cartographer_ros/cartographer_ros/ros_log_sink.cc:
```diff
@@ -20,9 +20,14 @@
 #include <cstring>
 #include <string>
 #include <thread>
+#include <iostream>
```
alphabetical

mikaelarguedas commented on cartographer_ros/cartographer_ros/ros_log_sink.cc:
```diff
@@ -20,9 +20,14 @@
 #include "glog/log_severity.h"
-#include "ros/console.h"
+
+#define ROS_INFO_STREAM(str) std::cout << str << std::endl;
```
I like it :+1:

mikaelarguedas commented on cartographer_ros/cartographer_ros/ros_log_sink.cc:
```diff
@@ -20,9 +20,14 @@
+#define ROS_INFO_STREAM(str) std::cout << str << std::endl;
+#define ROS_WARN_STREAM(str) std::cout << str << std::endl;
+#define ROS_ERROR_STREAM(str) std::cout << str << std::endl;
+#define ROS_FATAL_STREAM(str) std::cout << str << std::endl;
```
Should any of thos be printed to cerr ?

clalancette replied:
Hm, yeah, good point.  I'll probably make ERROR and FATAL cerr for now.

mikaelarguedas commented on cartographer_ros/cartographer_ros/time_conversion.cc:
```diff
@@ -17,31 +17,33 @@
 #include "cartographer/common/time.h"
-#include "ros/ros.h"
+#include <rclcpp/rclcpp.hpp>
+#include <builtin_interfaces/msg/time.hpp>
```
alphabetical order.
Consistency, let's pick `"` or `<` for all header from external packages and use the same everywhere

mikaelarguedas commented on cartographer_ros/cartographer_ros/urdf_reader.h:
```diff
@@ -21,10 +21,11 @@
 
 #include "cartographer/common/port.h"
 #include "tf2_ros/buffer.h"
+#include "geometry_msgs/msg/transform_stamped.hpp"
```
alphabetical

mikaelarguedas commented on cartographer_ros_msgs/CMakeLists.txt:
```diff
@@ -12,35 +12,35 @@
+find_package(builtin_interfaces REQUIRED)
+find_package(rosidl_default_generators REQUIRED)
+find_package(std_msgs REQUIRED)
+find_package(geometry_msgs REQUIRED)
```
alphabetical

mikaelarguedas commented on cartographer_ros/CMakeLists.txt:
```diff
@@ -57,43 +52,39 @@ if(DEFINED urdfdom_headers_VERSION)
-)
-
-file(GLOB_RECURSE ALL_SRCS "*.cc" "*.h")
-file(GLOB_RECURSE ALL_TESTS "*_test.cc")
```
can you either keep the existing tests or ticket it

clalancette commented on cartographer_ros/CMakeLists.txt:
```diff
@@ -12,42 +12,37 @@
+find_package(cartographer_ros_msgs REQUIRED)
+find_package(pcl_conversions REQUIRED)
+find_package(sensor_msgs REQUIRED)
+find_package(nav_msgs REQUIRED)
```
Absolutely.  Will fix.

clalancette commented on cartographer_ros/CMakeLists.txt:
```diff
@@ -57,43 +52,39 @@ if(DEFINED urdfdom_headers_VERSION)
 include_directories(
-  ${urdfdom_headers_INCLUDE_DIRS}
+  include
+  "."
```
Yeah, we need because a lot of the files do things like:

#include "cartographer_ros/msg_conversion.h"

clalancette commented on cartographer_ros/CMakeLists.txt:
```diff
@@ -57,43 +52,39 @@ if(DEFINED urdfdom_headers_VERSION)
+  ${cartographer_INCLUDE_DIRS}
+  ${cartographer_ros_msgs_INCLUDE_DIRS}
+  ${pcl_conversions_INCLUDE_DIRS}
+  ${nav_msgs_INCLUDE_DIRS}
```
Ha, I alphabetized that before, but just made a mistake.  Will fix.

clalancette commented on cartographer_ros/CMakeLists.txt:
```diff
@@ -105,72 +96,23 @@ foreach(DEFINITION ${PCL_DEFINITIONS})
-  LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
-  RUNTIME DESTINATION ${CATKIN_GLOBAL_BIN_DESTINATION}
+install(DIRECTORY configuration_files
+  DESTINATION ${CMAKE_INSTALL_PREFIX}/share/${PROJECT_NAME}/
```
Will fix.

clalancette commented on cartographer_ros/cartographer_ros/CMakeLists.txt:
```diff
@@ -12,35 +12,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-google_binary(cartographer_assets_writer
```
We can't currently port assets_writer to ROS2 because it needs an implementation of rosbag to function.  I can leave this commented out if preferred.

I'll do that for now with a TODO comment.

clalancette commented on cartographer_ros/cartographer_ros/assets_writer_main.cc:
```diff
@@ -150,7 +150,7 @@ void Run(const string& trajectory_filename, const string& bag_filename,
     bag.open(bag_filename, rosbag::bagmode::Read);
     rosbag::View view(bag);
-    const ::ros::Time begin_time = view.getBeginTime();
+    const ::builtin_interfaces::msg::Time begin_time = view.getBeginTime();
```
Actually, this particular file is an interesting case.  The changes that were done here are only because I did various 'sed' commands across all C++ files.  As-is, this has no chance of working because we need rosbag for this file.  So we could either revert back to the original, and just know it won't work completely, or we could leave some of these changes.  What's preferred?

mikaelarguedas replied:
Yeah I agree that these files can not be ported given that they rely on missing features. It's just surprising to have a sed on "ros::Time" but not the corresponding one on "ros/time.h". I think they should go together, either both changes or none of them.

clalancette replied:
Yeah, that's fine.  I'll actually revert this back to the original, and then we'll just have to port the whole thing in the future.

clalancette commented on cartographer_ros/cartographer_ros/map_builder_bridge.h:
```diff
@@ -25,10 +25,11 @@
+#include "cartographer_ros_msgs/msg/submap_list.hpp"
+#include "cartographer_ros_msgs/srv/submap_query.hpp"
+#include "nav_msgs/msg/occupancy_grid.hpp"
+#include <rclcpp/rclcpp.hpp>
```
I mostly just kept the quote vs bracket distinction as they were in the original cartographer code.  That being said, I think we prefer brackets for ROS2 headers, so I'll change those (this won't make any difference for merging from upstream, since I'm changing those lines anyway).

clalancette commented on cartographer_ros/cartographer_ros/node.cc:
```diff
@@ -62,55 +66,50 @@ Node::~Node() {
-        node_handle_.advertise<::nav_msgs::OccupancyGrid>(
-            kOccupancyGridTopic, kLatestOnlyPublisherQueueSize,
-            true /* latched */);
+      node_handle_->create_publisher<::nav_msgs::msg::OccupancyGrid>(kOccupancyGridTopic, rmw_qos_profile_default);
```
Yeah, very good point.  I actually haven't gotten these through successfully yet, this may be one reason.  I'll fix this.

clalancette commented:
CI:
turtlebot amd64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux&build=16)](http://ci.ros2.org/job/ci_turtlebot-demo_linux/16/)
turtlebot aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux-aarch64&build=32)](http://ci.ros2.org/job/ci_turtlebot-demo_linux-aarch64/32/)

clalancette commented:
So, I ended up rebasing this on the latest cartographer_ros code, which caused a few things to move around.  Overall, though, the code is pretty much the same as has been reviewed before.  I'm going to run another CI on it now.

CI:
turtlebot amd64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux&build=25)](http://ci.ros2.org/job/ci_turtlebot-demo_linux/25/)
turtlebot aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux-aarch64&build=41)](http://ci.ros2.org/job/ci_turtlebot-demo_linux-aarch64/41/)

clalancette commented:
I think I have all review comments addressed, I'm on the latest code, CI is green, and @wjwwood approved.  If there are no further comments, I'll merge this a little later on.

mikaelarguedas commented:
If you think that the recent changes don't need review :+1: for merging it. Can you please rebase your branch to not include all the upstream commits before merging ? that'll make the diff more readable and the PR mor self contained.

clalancette commented:
The weird thing is that I did that already.  I'm not sure why github is still showing all of those differences.  I'll try rebasing again (there is one new commit that shouldn't affect us), and see if that fixes it.

mikaelarguedas commented:
oh I had the same problem yesterday just before github went down, I had to close and reopen the PR for the commit list and the diff to update :confounded:

clalancette commented:
There we go, that's better.  The merge looks a lot more reasonable now.  I'm going to merge this up.
